test/src/unit-conversions.cpp

```diff
@@ -1692,3 +1700,74 @@ TEST_CASE("JSON to enum mapping")
         CHECK(TS_INVALID == json("what?").get<TaskState>());
     }
 }
+
+#ifdef JSON_HAS_CPP_17
+TEST_CASE("std::optional")
+{
+    SECTION("null")
+    {
+        json j_null;
+        std::optional<std::string> opt_null;
+
+        CHECK(json(opt_null) == j_null);
+        CHECK_THROWS_WITH(std::optional<std::string>(j_null),
```

Comment:
Weird. In my local env, the exception would not raise. By from CI log, it will get the exception.

Reply:
What is special in your local environment?

Reply:
ubuntu 18.04
gcc 7.4
cmake 3.10.2

Reply:
Somewhat has changed since GCC 7.4
https://godbolt.org/z/4MfL8i

Reply:
There is probably a bug in GCC 7. When used with "-std=c++17" it may perform wrong overload resolution for user-defined conversion. In previous example and also [here](https://godbolt.org/z/kWTWx9) (with no `std::optional` involved) it behaves as if the implicit copy constructor is selected, whereas there is a more suitable template constructor. Other major releases of GCC (including older versions) prefer template constructor.